### DISCUSSION TITLE
Produktivserver

### DISCUSSION THREAD

**Discussion by pbroska** [Digitaler-Impfnachweis/certification-apis#103] [Q&A] [answered]
Hallo,
sind die Produktivserver für die Ärzte bereits freigeschaltet, oder können wir nur in der Referenzumgebung testen?
MfG
Patrick

**Comment by ManuelB** [comment, marked as answer]
@pbroska Produktivsysteme sollten gehen. RU sollte auch gehen.

Im Produktivsystem sollte man nur mit echten Daten "testen".

**Reply by pbroska** [reply]
Danke für die Antwort. Leider kriegen wir keinen Token vom Produktivserver (internal server error). Ist da irgendwas bekannt?
https://id.impfnachweis.info/auth/realms/bmg-ti-certify/protocol/openid-connect/token?grant_type=authorization_code&redirect_uri=connector%3A%2F%2Fauthenticated&client_id=cert-auth&session_state=b99ada54-b452-4904-b94c-0bbbb4d127a4&code=3b9895...

Der Testkonnektor für die Referenzumgebung kommt heute erst an, deswegen kann ich noch nicht in der Referenzumgebung testen.

**Reply by ManuelB** [reply]
@pbroska ich würde sagen deine client_id ist falsch. Die müsste: user-access-ti sein.

**Reply by pbroska** [reply]
Leider keine Änderung, session_state und code gerade neu erzeugt:
https://id.impfnachweis.info/auth/realms/bmg-ti-certify/protocol/openid-connect/token?grant_type=authorization_code&redirect_uri=connector%3A%2F%2Fauthenticated&**client_id=user-access-ti**&session_state=19e7ee5a-6c9f-4813-af98-3b54f5aacaad&code=178047c4-08f5-48ea-b77d-987ba5ae03ad.19e7ee5a-6c9f-4813-af98-3b54f5aacaad.29638cf4-d546-49e0-9945-29bc0a4e3ed6

**Reply by pbroska** [reply]
Die beiden Sternchen sollte Fett auslösen, habe ich nachträglich hinzugefügt. Ist also kein Bestandteil des strings.

**Reply by mge-medisoftware** [reply]
Sieht eigentlich gut aus. Wir übertragen die Parameter allerdings im body per POST mit "Content-Type: application/x-www-form-urlencoded"

**Reply by pbroska** [reply]
Umgestellt von Header auf body, jetzt scheint nur noch der Login nicht zu klappen:
{"error":"unauthorized_client","error_description":"INVALID_CREDENTIALS: Invalid client credentials"}* Connection #1 to host keycloak-cert-auth.dev.ibmega.net left intact
Gibt es bei der Signierung etwas zu beachten? Ich verwende C.AUT.

**Reply by pbroska** [reply]
Umgestellt auf Testumgebung ohne Konnektor und client_id=smcb_auth_simulator, aber mit Fehler:
curl POST --url "https://keycloak-cert-auth.dev.ibmega.net/auth/realms/master/protocol/openid-connect/token" -d "grant_type=authorization_code&client_id=smcb-auth-simulator&redirect_uri=connector://authenticated?session_state=a75f4fff-7717-42df-a3c9-c8e5551a0db2&code=b716612f-4af8-4f3f-b1e3-d42626b7a5c5.a75f4fff-7717-42df-a3c9-c8e5551a0db2.d5af5f96-dc70-4402-add3-954386b639f9" -v

{"error":"invalid_grant","error_description":"Code not valid"}* Connection #1 to host keycloak-cert-auth.dev.ibmega.net left intact

**Reply by mge-medisoftware** [reply]
Die challenge (ohne irgendeine Modifikation direkt als String) hashen mit SHA256. Den Hash Base64 enkodieren. Damit ExternalAuthenticate() mit Signaturtyp PKCS1 ausführen. Zertifikat C.AUT ist korrekt (mit secunet-Konnektor getestet). KoPS/Puppetry haben wir nicht verwendet, da zu unserem Implementierungszeitpunkt  die damalige Version ExternalAuthenticate()  noch nicht unterstützte.
Ist halt die Frage was serverseitig geprüft wird, gibt/gab ja wohl auch ein BSNR-Whitelisting oder ein OID-Check zum Zertifikat.

**Reply by pbroska** [reply]
Endlich, meine erster TOKEN!!!
Es lag an der curl Version von Windows 10. Mit der aktuellen Version 7.77.0 64Bit gab es keine Probleme mehr.
Danke an alle Replies!!!
